src/main/cpp/src/hyper_log_log_plus_plus.cu

```diff
@@ -663,9 +664,13 @@ template <int block_size>
 __launch_bounds__(block_size) CUDF_KERNEL
   void reduce_hllpp_kernel(cudf::column_device_view hashs,
                            cudf::device_span<int64_t*> output,
-                           int precision)
+                           int precision,
+                           int32_t* mem_cache)
 {
-  __shared__ int32_t shared_data[block_size];
+  extern __shared__ int32_t shared_mem_cache[];
+
+  int32_t* cache = shared_mem_cache;
+  if (precision > 12) { cache = mem_cache; }
```

Comment:
Instead of hardcoding `12` in several places, please use a global variable, for example: `constexpr int MAX_SHARED_MEM_PRECISION = 12`.